server/src/main/java/io/crate/planner/optimizer/symbol/rule/SwapCastsInComparisonOperators.java

```diff
@@ -58,11 +58,6 @@ private static boolean isSwappable(Function cmpOp) {
         if (cmpOp.arguments().get(1) instanceof Literal<?> literal) {
             Function cast = (Function) cmpOp.arguments().getFirst();
             Reference ref = (Reference) cast.arguments().getFirst();
-            DataType<?> refInnerType = ArrayType.unnest(ref.valueType());
-            DataType<?> literalInnerType = ArrayType.unnest(literal.valueType());
-            if (!DataTypes.isNumeric(literalInnerType) || !DataTypes.isNumeric(refInnerType)) {
-                return true;
-            }
```

Comment:
x = true is a valid scenario, basically any non-null/non-zero-bytes value must be hit.

On master we mistakenly throw ClassCastException.

I probably could add more checks to   
`if (!DataTypes.isNumeric(literalInnerType) || !DataTypes.isNumeric(refInnerType)) {`

but since `isNarrowingConversionPossible` is anyways relies on try-catch, let's just rely on it for all cases to reduce complexity and avoid error-prone "avoid-try-catch" optimization

Reply:
I was wrong, we can't use `isNarrowingConversionPossible` for non-numeric types but we also can't just return true for other pairs

Solution: use number specific logic for numbers like it was before but actually check casting for other types